src/osd/PrimaryLogPG.cc

```diff
@@ -6724,10 +6725,16 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 
 	if (op.extent.length == 0) {
 	  if (op.extent.offset > oi.size) {
-	    t->truncate(
-	      soid, op.extent.offset);
-            truncate_update_size_and_usage(ctx->delta_stats, oi,
-                                           op.extent.offset);
+	    if (seq && (seq > op.extent.truncate_seq)) {
+	      //do nothing
+	      //write arrived after truncate, we should not truncate to offset
+	    } else {
+	      t->truncate(
+	        soid, op.extent.offset);
+	      truncate_update_size_and_usage(ctx->delta_stats, oi,
+	                                     op.extent.offset);
+	      oi.clear_data_digest();
+	    }
```

Comment:
I don't think this part is correct.  The call to t->truncate here I think has nothing to do with the delayed truncate behavior -- it's to logically zero pad the object out to the offset of the new write.

Also, the block immediately below this one should already be calling oi.clear_data_digest().